CMakeLists.txt

```diff
@@ -26,6 +20,16 @@ OPTION(BUILD_STATIC_LIB "Build static library" OFF)
 OPTION(__BUILD_FOR_R "Set to ON if building lib_lightgbm for use with the R package" OFF)
 OPTION(__INTEGRATE_OPENCL "Set to ON if building LightGBM with the OpenCL ICD Loader and its dependencies included" OFF)
 
+if(__BUILD_FOR_R)
+  set(CMAKE_TRY_COMPILE_TARGET_TYPE "STATIC_LIBRARY")
+endif(__BUILD_FOR_R)
+
+if(USE_CUDA)
```

Comment:
I moved this down here because `set(CMAKE_TRY_COMPILE_TARGET_TYPE)` has to run before `PROJECT()`, and I needed to access the result of `OPTION(__BUILD_FOR_R)` for it